lib/Sema/CSDiagnostics.cpp

```diff
@@ -984,37 +984,76 @@ bool MissingExplicitConversionFailure::diagnoseAsError() {
+    if (componentPathElt->getIndex() == 0) {
+      if (auto rootType = keyPathExpr->getRootType()) {
+        return rootType->getSourceRange();
+      }
+    } else {
+      auto componentIdx = componentPathElt->getIndex() - 1;
+      auto component = keyPathExpr->getComponents()[componentIdx];
+      return component.getSourceRange();
+    }
+  }
+  return FailureDiagnostic::getSourceRange();
+}
+
 bool MemberAccessOnOptionalBaseFailure::diagnoseAsError() {
-  auto anchor = getAnchor();
-  auto baseType = getType(anchor);
+  auto baseType = getMemberBaseType();
+  auto locator = getLocator();
+  
   bool resultIsOptional = ResultTypeIsOptional;
 
   // If we've resolved the member overload to one that returns an optional
   // type, then the result of the expression is optional (and we want to offer
   // only a '?' fixit) even though the constraint system didn't need to add any
   // additional optionality.
-  auto overload = getOverloadChoiceIfAvailable(getLocator());
+  auto overload = getOverloadChoiceIfAvailable(locator);
   if (overload && overload->openedType->getOptionalObjectType())
     resultIsOptional = true;
 
   auto unwrappedBaseType = baseType->getOptionalObjectType();
   if (!unwrappedBaseType)
     return false;
-
-  emitDiagnostic(diag::optional_base_not_unwrapped, baseType, Member,
-                 unwrappedBaseType);
-
-  // FIXME: It would be nice to immediately offer "base?.member ?? defaultValue"
-  // for non-optional results where that would be appropriate. For the moment
-  // always offering "?" means that if the user chooses chaining, we'll end up
-  // in MissingOptionalUnwrapFailure:diagnose() to offer a default value during
-  // the next compile.
-  emitDiagnostic(diag::optional_base_chain, Member)
-      .fixItInsertAfter(getSourceRange().End, "?");
-
-  if (!resultIsOptional) {
-    emitDiagnostic(diag::unwrap_with_force_value)
-        .fixItInsertAfter(getSourceRange().End, "!");
+  
+  auto sourceRange = getSourceRange();
+  
+  emitDiagnostic(diag::optional_base_not_unwrapped,
+                 baseType, Member, unwrappedBaseType);
+
+  auto componentPathElt =
+      locator->getLastElementAs<LocatorPathElt::KeyPathComponent>();
+  if (componentPathElt && componentPathElt->getIndex() == 0) {
+    // For members where the base type is an optional key path root
+    // let's emit a tailored note suggesting to remove the '?' and a
+    // fix to replace the optional type with its unwrapped type.
+    auto *keyPathExpr = castToExpr<KeyPathExpr>(getAnchor());
+    if (auto rootType = keyPathExpr->getRootType()) {
+      emitDiagnostic(diag::optional_base_remove_optional_for_keypath_root,
+                     unwrappedBaseType)
+          .fixItReplace(rootType->getSourceRange(),
+                        unwrappedBaseType.getString());
+    } else {
+      emitDiagnostic(diag::optional_base_chain, Member)
+          .fixItInsertAfter(keyPathExpr->getLoc(), "?");
```

Comment:
I'm not sure this suggestion is correct, AFAIK we don't support optional chaining on the root of a key path.

Reply:
Ah true, so I think we can't really suggest any fix here ... :( 
I would think about inserting the unwrapped type here but I just check it gives 
```swift
func SR5688_KP(_ kp: KeyPath<String?, Int>) {}

SR5688_KP(\String.count) // Key path value type 'String?' cannot be converted to contextual type 'String'
```
Which I don't think is correct, or it is at the very least a bad diagnostic ... not sure

Reply:
Just remove it, not really sure if we could suggest something else here ... 
cc @xedin

Reply:
@hamishknight :)
I was thinking more about this today and to me, maybe misunderstanding the diagnostics, but this still feels like a problem. I'm not really sure if that is something that supposed to work or if it is a matter of improving diagnostics for this case. What do you think? Should we open a ticket to track this if it is indeed a problem?

Reply:
Yeah, I agree it could be better – perhaps we should have a custom diagnostic that emphasises that the problem is with the key path's root type being inferred to an optional type? Ideally we'd be able to point to where the inferred type is coming from in the user's code, but we don't yet track that information in the constraint system.

Reply:
> Yeah, I agree it could be better – perhaps we should have a custom diagnostic that emphasises that the problem is with the key path's root type being inferred to an optional type?

I think we could do that, as you mentioned the ideal would be to point out the declaration and suggest not use optional as keypath root. I'll adjust the message asap then :)

Reply:
@hamishknight I just added a TODO diagnostics for this, it is ok or do you think it is something that should be handled here on this patch?

Reply:
@LucianoPAlmeida I'm totally fine with leaving it as a TODO for now – would you mind also filing it as a bug/improvement?

Reply:
Yeap, Done https://bugs.swift.org/browse/SR-13079 :)